Union return values with generics from queryset methods resolve wrong when the queryset is used on multiple models.

# Bug report

<!--
Hi, thanks for submitting a bug. We appreciate that.

But, we will need some information about what's wrong to help you.
-->

## What's wrong

Returning a union type that includes a generic from a queryset method results with the generic getting resolved wrong on subsequent models.

```
-   case: handles_type_var_in_subclasses_of_subclasses_of_queryset
    main: |
        from myapp.models import MyModel, MyOtherModel
        reveal_type(MyModel.objects.example_other_mixin())  # N: Revealed type is "myapp.models.MyModel"
        reveal_type(MyModel.objects.example_union())  # N: Revealed type is "Union[myapp.models.MyModel, None]"
        reveal_type(MyOtherModel.objects.example_other_mixin())  # N: Revealed type is "myapp.models.MyOtherModel"
        reveal_type(MyOtherModel.objects.example_union())  # N: Revealed type is "Union[myapp.models.MyOtherModel, None]"
    installed_apps:
        - myapp
    files:
        -   path: myapp/__init__.py
        -   path: myapp/models.py
            content: |
              from typing import TypeVar, Generic, Optional
              from django.db import models
              from typing_extensions import Self

              T = TypeVar("T", bound=models.Model)
              T_2 = TypeVar("T_2", bound=models.Model)

              class SomeMixin:
                  def example_mixin(self, a: T) -> T: ...

              class OtherMixin(models.QuerySet[T]):
                  def example_other_mixin(self) -> T: ...
                  def example_union(self) -> Optional[T]: ...

              class _MyModelQuerySet(OtherMixin[T], models.QuerySet[T], Generic[T]):
                  def example(self) -> T: ...
                  def override(self) -> T: ...
                  def override2(self) -> T: ...
                  def dummy_override(self) -> int: ...
                  def test_sub_self(self) -> Self: ...

              class _MyModelQuerySet2(SomeMixin, _MyModelQuerySet[T_2]):
                  def example_2(self) -> T_2: ...
                  def override(self) -> T_2: ...
                  def override2(self) -> T_2: ...
                  def dummy_override(self) -> T_2: ...  # type: ignore[override]
                  def test_self(self) -> Self: ...

              class MyModelQuerySet(_MyModelQuerySet2["MyModel"]):
                  def override(self) -> "MyModel": ...

              class MyModel(models.Model):
                  objects = MyModelQuerySet.as_manager()

              class MyOtherModel(models.Model):
                  objects = _MyModelQuerySet2.as_manager()
```

Result:
both `reveal_type(MyModel.objects.example_union())` and `reveal_type(MyOtherModel.objects.example_union())` get resolved to `"Union[myapp.models.MyModel, None]"`
```
============================================================================================ FAILURES ============================================================================================
____________________________________________________________________ handles_type_var_in_subclasses_of_subclasses_of_queryset ____________________________________________________________________
/Users/jkaikko/Projects/django-stubs/tests/typecheck/managers/querysets/test_as_manager.yml:249:
E   pytest_mypy_plugins.utils.TypecheckAssertionError: Invalid output:
E   Actual:
E     ...
E     main:20: note: Revealed type is "Union[myapp.models.MyModel, None]" (diff)
E   Expected:
E     ...
E     main:20: note: Revealed type is "Union[myapp.models.MyOtherModel, None]" (diff)
E   Alignment of first line difference:
E     E: ...s "Union[myapp.models.MyOtherModel, None]"
E     A: ...s "Union[myapp.models.MyModel, None]"
E                                   ^
```
## How is that should be

`reveal_type(MyModel.objects.example_union())` should be resolve to `Union[myapp.models.MyModel, None]` and `MyOtherModel.objects.example_union()` should resolve to `Union[myapp.models.MyOtherModel, None]`

## System information

- OS:
- `python` version: 3.9.9, 3.11.9
- `django` version: 4.2, 5.0
- `mypy` version: 1.11.1
- `django-stubs` version: 5.0.4
- `django-stubs-ext` version: 5.0.4

--- a/mypy_django_plugin/transformers/managers.py
+++ b/mypy_django_plugin/transformers/managers.py
@@ -1,6 +1,7 @@
 from typing import Final, Optional, Union
 
 from mypy.checker import TypeChecker
+from mypy.copytype import copy_type
 from mypy.nodes import (
     GDEF,
     CallExpr,
@@ -244,15 +245,18 @@ def _replace_type_var(ret_type: MypyType, to_replace: str, replace_by: MypyType)
     """
     if isinstance(ret_type, TypeVarType) and ret_type.fullname == to_replace:
         return replace_by
-    elif isinstance((instance := get_proper_type(ret_type)), Instance):
-        # Since it is an instance, recursively find the type var for all its args.
-        instance.args = tuple(_replace_type_var(item, to_replace, replace_by) for item in instance.args)
-        return instance
 
-    if isinstance(ret_type, ProperType) and hasattr(ret_type, "item"):
+    ret_type = copy_type(get_proper_type(ret_type))
+
+    if isinstance(ret_type, Instance):
+        # Since it is an instance, recursively find the type var for all its args.
+        ret_type.args = tuple(_replace_type_var(item, to_replace, replace_by) for item in ret_type.args)
+        return ret_type
+
+    if hasattr(ret_type, "item"):
         # For example TypeType has an item. find the type_var for this item
         ret_type.item = _replace_type_var(ret_type.item, to_replace, replace_by)
-    if isinstance(ret_type, ProperType) and hasattr(ret_type, "items"):
+    if hasattr(ret_type, "items"):
         # For example TypeList has items. find recursively type_var for its items
         ret_type.items = [_replace_type_var(item, to_replace, replace_by) for item in ret_type.items]
     return ret_type

--- a/tests/typecheck/managers/querysets/test_as_manager.yml
+++ b/tests/typecheck/managers/querysets/test_as_manager.yml
@@ -236,6 +236,7 @@
         reveal_type(MyModel.objects.dummy_override())  # N: Revealed type is "myapp.models.MyModel"
         reveal_type(MyModel.objects.example_mixin(MyModel()))  # N: Revealed type is "myapp.models.MyModel"
         reveal_type(MyModel.objects.example_other_mixin())  # N: Revealed type is "myapp.models.MyModel"
+        reveal_type(MyModel.objects.example_union())  # N: Revealed type is "Union[myapp.models.MyModel, builtins.list[myapp.models.MyModel]]"
         reveal_type(MyOtherModel.objects.example())  # N: Revealed type is "myapp.models.MyOtherModel"
         reveal_type(MyOtherModel.objects.example_2())  # N: Revealed type is "myapp.models.MyOtherModel"
         reveal_type(MyOtherModel.objects.override())  # N: Revealed type is "myapp.models.MyOtherModel"
@@ -245,13 +246,14 @@
         reveal_type(MyOtherModel.objects.example_other_mixin())  # N: Revealed type is "myapp.models.MyOtherModel"
         reveal_type(MyOtherModel.objects.test_self())  # N: Revealed type is "myapp.models._MyModelQuerySet2[myapp.models.MyOtherModel]"
         reveal_type(MyOtherModel.objects.test_sub_self())  # N: Revealed type is "myapp.models._MyModelQuerySet2[myapp.models.MyOtherModel]"
+        reveal_type(MyOtherModel.objects.example_union())  # N: Revealed type is "Union[myapp.models.MyOtherModel, builtins.list[myapp.models.MyOtherModel]]"
     installed_apps:
         - myapp
     files:
         -   path: myapp/__init__.py
         -   path: myapp/models.py
             content: |
-              from typing import TypeVar, Generic
+              from typing import TypeVar, Generic, Union, List
               from django.db import models
               from typing_extensions import Self
 
@@ -263,6 +265,7 @@
 
               class OtherMixin(models.QuerySet[T]):
                   def example_other_mixin(self) -> T: ...
+                  def example_union(self) -> Union[T, List[T]]: ...
 
               class _MyModelQuerySet(OtherMixin[T], models.QuerySet[T], Generic[T]):
                   def example(self) -> T: ...

